proxy.h

```diff
@@ -121,41 +110,91 @@ template <class R, class... Args>
+  template <class D>
+  struct meta_provider {
+    template <class P>
+    static R dispatcher(const char* erased, Args... args) {
+      auto ptr = ptr_traits<P>::to_address(*reinterpret_cast<const P*>(erased));
+      if constexpr (std::is_void_v<R>) {
+        D{}(*ptr, std::forward<Args>(args)...);
+      } else {
+        return D{}(*ptr, std::forward<Args>(args)...);
+      }
+    }
+  };
 
+  template <class F, class T>
+  static constexpr bool applicable_callable =
+      std::is_invocable_r_v<R, F, T&, Args...>;
   template <class D, class P>
-  static constexpr bool applicable_ptr = std::is_invocable_v<
-      D, typename ptr_traits<P>::reference_type, Args...>;
+  static constexpr bool applicable_ptr =
+      applicable_callable<D, typename ptr_traits<P>::reference_type>;
   static constexpr bool is_noexcept = false;
-  template <class D, class P>
-  static R dispatcher(const char* erased, Args... args) {
-    auto ptr = ptr_traits<P>::to_address(*reinterpret_cast<const P*>(erased));
-    if constexpr (std::is_void_v<R>) {
-      D{}(*ptr, std::forward<Args>(args)...);
-    } else {
-      return D{}(*ptr, std::forward<Args>(args)...);
-    }
-  }
 };
 template <class R, class... Args>
 struct overload_traits<R(Args...) noexcept> : applicable_traits {
   using dispatcher_type = R (*)(const char*, Args...) noexcept;
   struct resolver { R (*operator()(Args...))(Args...) noexcept; };
-  using forwarding_argument_types = std::tuple<Args&&...>;  // For helper macros
+  using forwarding_argument_types = std::tuple<Args&&...>;
+  template <class D>
+  struct meta_provider {
+    template <class P>
+    static R dispatcher(const char* erased, Args... args) noexcept {
+      auto ptr = ptr_traits<P>::to_address(*reinterpret_cast<const P*>(erased));
+      if constexpr (std::is_void_v<R>) {
+        D{}(*ptr, std::forward<Args>(args)...);
+      } else {
+        return D{}(*ptr, std::forward<Args>(args)...);
+      }
+    }
+  };
 
+  template <class F, class T>
+  static constexpr bool applicable_callable =
+      std::is_nothrow_invocable_r_v<R, F, T&, Args...>;
   template <class D, class P>
-  static constexpr bool applicable_ptr = std::is_nothrow_invocable_v<
-      D, typename ptr_traits<P>::reference_type, Args...>;
+  static constexpr bool applicable_ptr =
+      applicable_callable<D, typename ptr_traits<P>::reference_type>;
   static constexpr bool is_noexcept = true;
-  template <class D, class P>
-  static R dispatcher(const char* erased, Args... args) noexcept {
-    auto ptr = ptr_traits<P>::to_address(*reinterpret_cast<const P*>(erased));
-    if constexpr (std::is_void_v<R>) {
-      D{}(*ptr, std::forward<Args>(args)...);
-    } else {
-      return D{}(*ptr, std::forward<Args>(args)...);
-    }
-  }
+};
+
+template <class T> struct nullable_traits : inapplicable_traits {};
+template <class T>
+    requires(
+        requires(const T& cv, T& v) {
+          T{};
+          { cv.has_value() } -> std::convertible_to<bool>;
```

Comment:
Add the noexcept constraint to `has_value()` and `reset()`?

Reply:
Added